ttl_test.go

```diff
@@ -174,6 +176,121 @@ func TestTTLScenario(t *testing.T) {
 	}
 }
 
+func TestHeadTailLogicConcurrent(t *testing.T) {
+	m := NewMapTTLCache[string, string](context.Background(), time.Millisecond, time.Hour)
+
+	pool := make([]string, 50)
+	for i := range pool {
+		pool[i] = randomString(10)
+	}
+
+	wg := sync.WaitGroup{}
+	for i := 0; i < 1000; i++ {
+		idx := rand.Intn(len(pool))
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+
+			_, err := m.Get(pool[idx])
+			if err != nil {
+				m.Set(pool[idx], pool[idx])
+			}
+		}()
+	}
+
+	wg.Wait()
+	time.Sleep(time.Millisecond * 3)
+
+	var expectedHeadKey, expectedTailKey string
+	var expectedHeadValue, expectedTailValue ttlRec[string, string]
+
+	keys := map[string]struct{}{}
+
+	for k, v := range m.data {
+		keys[k] = struct{}{}
+		if expectedTailKey == "" {
+			expectedTailKey = k
+			expectedHeadKey = k
+			expectedHeadValue = v
+			expectedTailValue = v
+		} else if expectedTailValue.timestamp.Before(v.timestamp) {
+			expectedTailKey = k
+			expectedTailValue = v
+		} else if expectedHeadValue.timestamp.After(v.timestamp) {
+			expectedHeadKey = k
+			expectedHeadValue = v
+		}
+	}
+
+	for k, v := range m.data {
+		if _, ok := keys[v.next]; k != m.tail && !ok {
+			t.Errorf("expected key %q not found in data", v.next)
+		}
+
+		if _, ok := keys[v.prev]; k != m.head && !ok {
+			t.Errorf("expected key %q not found in data", v.prev)
+		}
+	}
+
+	if m.head != expectedHeadKey {
+		t.Errorf("expected head key %q, but got %v", expectedHeadKey, m.head)
+	}
+
+	if m.tail != expectedTailKey {
+		t.Errorf("expected tail key %q, but got %v", expectedTailKey, m.tail)
+	}
+
+	if err := m.cleanup(); err != nil {
+		t.Errorf("unexpected error in cleanup: %v", err)
+	}
+
+	if m.Len() != 0 {
```

Comment:
Should we also add a check that head and tail are zero values? Just in case.

Reply:
added

Reply:
Oh now it found coverage decrease. I can add missing tests later.